iroh-gateway/src/core.rs

```diff
@@ -385,40 +366,143 @@ async fn serve_car(
+    match metadata.unixfs_type {
+        Some(UnixfsType::Dir) => {
+            if let Some(dir_list_data) = body.data().await {
+                let dir_list = match dir_list_data {
+                    Ok(b) => b,
+                    Err(_) => {
+                        return Err(error(
+                            StatusCode::INTERNAL_SERVER_ERROR,
+                            "failed to read dir listing",
+                            &state,
+                        ));
+                    }
+                };
+                return serve_fs_dir(&dir_list, req, state, headers, start_time).await;
+            } else {
+                return Err(error(
+                    StatusCode::INTERNAL_SERVER_ERROR,
+                    "failed to read dir listing",
+                    &state,
+                ));
+            }
+        }
+        Some(_) => {
+            // todo(arqu): error on no size
+            // todo(arqu): add lazy seeking
+            add_cache_control_headers(&mut headers, metadata.clone());
+            set_etag_headers(&mut headers, get_etag(&req.cid, Some(req.format.clone())));
+            let name = add_content_disposition_headers(
+                &mut headers,
+                &req.query_file_name,
+                &req.content_path,
+                req.download,
+            );
+            if metadata.unixfs_type == Some(UnixfsType::Symlink) {
+                headers.insert(
+                    CONTENT_TYPE,
+                    HeaderValue::from_str("inode/symlink").unwrap(),
+                );
+            } else {
+                add_content_type_headers(&mut headers, &name);
+            }
+        }
+        None => {
+            return Err(error(
+                StatusCode::BAD_REQUEST,
+                "couldn't determine unixfs type",
+                &state,
+            ));
+        }
+    }
     response(StatusCode::OK, body, headers)
 }
 
+#[tracing::instrument()]
+async fn serve_fs_dir(
+    dir_list: &Bytes,
+    req: &Request,
+    state: Arc<State>,
+    mut headers: HeaderMap,
+    start_time: std::time::Instant,
+) -> Result<GatewayResponse, GatewayError> {
+    let dir_list = String::from_utf8(dir_list.to_vec()).unwrap();
+    let dir_list_lines = dir_list.lines();
+    let force_dir = req.query_params.force_dir.unwrap_or(false);
+    if !force_dir {
+        for line in dir_list_lines.clone() {
+            if line == "index.html" {
+                if !req.content_path.ends_with('/') {
+                    let redirect_path = format!(
+                        "{}/{}",
+                        req.content_path,
+                        req.query_params.to_query_string()
+                    );
+                    return Ok(GatewayResponse::redirect(&redirect_path));
+                }
+                let mut new_req = req.clone();
+                new_req
+                    .resolved_path
+                    .extend_tail(vec!["index.html".to_string()]);
```

Comment:
maybe sth like `fn push(&mut self, str: impl AsRef<str>)` would make for a nicer api